dpctl/tensor/_array_api.py

```diff
@@ -0,0 +1,118 @@
+
+
+def _isdtype_impl(dtype, kind):
+    if isinstance(kind, dpt.dtype):
+        return dtype == kind
+
+    elif isinstance(kind, str):
+        if kind == "bool":
+            return dtype.kind == "b"
+        elif kind == "signed integer":
+            return dtype.kind == "i"
+        elif kind == "unsigned integer":
+            return dtype.kind == "u"
+        elif kind == "integral":
+            return dtype.kind in "iu"
+        elif kind == "real floating":
+            return dtype.kind == "f"
+        elif kind == "complex floating":
+            return dtype.kind == "c"
+        elif kind == "numeric":
+            return dtype.kind in "iufc"
+        else:
+            raise ValueError(f"Unrecognized data type kind: {kind}")
+
+    elif isinstance(kind, tuple):
+        return any(_isdtype_impl(dtype, k) for k in kind)
+    else:
+        raise TypeError(f"Unsupported data type kind: {kind}")
+
+
+class __array_namespace_info__:
+    def __init__(self):
+        self._capabilities = {
+            "boolean_indexing": True,
+            "data_dependent_shapes": True,
+        }
+        self._all_dtypes = {
+            "bool": dpt.bool,
+            "float16": dpt.float16,
+            "float32": dpt.float32,
+            "float64": dpt.float64,
+            "complex64": dpt.complex64,
+            "complex128": dpt.complex128,
+            "int8": dpt.int8,
+            "int16": dpt.int16,
+            "int32": dpt.int32,
+            "int64": dpt.int64,
+            "uint8": dpt.uint8,
+            "uint16": dpt.uint16,
+            "uint32": dpt.uint32,
+            "uint64": dpt.uint64,
+        }
+
+    def capabilities(self):
+        return self._capabilities.copy()
+
+    def default_device(self):
+        return dpctl.select_default_device()
+
+    def default_dtypes(self, device=None):
+        if device is None:
+            device = dpctl.select_default_device()
+        return {
+            "real floating": default_device_fp_type(device),
+            "complex floating": default_device_complex_type,
+            "integral": default_device_int_type(device),
+            "indexing": default_device_index_type(device),
+        }
+
+    def dtypes(self, device=None, kind=None):
+        if device is None:
+            device = dpctl.select_default_device()
+        ignored_types = []
+        if not device.has_aspect_fp16:
+            ignored_types.append("float16")
+        if not device.has_aspect_fp64:
+            ignored_types.append("float64")
+        if kind is None:
+            return {
+                key: val
+                for key, val in self._all_dtypes.items()
+                if key not in ignored_types
+            }
+        else:
+            return {
+                key: val
+                for key, val in self._all_dtypes.items()
+                if key not in ignored_types and _isdtype_impl(val, kind)
+            }
+
+    def devices(self):
+        return dpctl.get_devices()
```

Comment:
It is possible to run in an environment where CUDA device is available, but `dpctl` was not compiled to offload to NVidia GPUs. 

Perhaps we should thin this list by ensuring that we can create an `dpt.arange` for this device and cache that list. 

It does not need to be done in this PR though